lib/trie/database.go

```diff
@@ -185,22 +185,28 @@ func (t *Trie) loadNode(db Database, n *Node) error {
 	return nil
 }
 
-// PopulateMerkleValues writes the Merkle value of each children of the node given
-// as keys to the map merkleValues.
-func (t *Trie) PopulateMerkleValues(n *Node, merkleValues map[string]struct{}) {
-	if n.Kind() != node.Branch {
+// PopulateMerkleValues writes the Merkle values of the node given and of
+// all its descendant nodes as keys to the map merkleValues.
+// It is assumed the node and its descendant nodes have their Merkle value already
```

Comment:
Is this assumption reenforced elsewhere in the code?

Reply:
Good point;

If this function was previously working and ignoring non-pre-calculated Merkle values, it was just plain wrong and dangerous, so we should panic.

Now practically speaking:

1. This is only used in the offline pruner in this particular git branch, so it's really not-a-big-deal, and it would work fine (due to LoadFromDB caching Merkle values for each trie node loaded)
2. #2873 (based on this branch) uses `PopulateMerkleValues` within the trie code to track deleted merkle values, so it becomes more critical. That's why I added https://github.com/ChainSafe/gossamer/pull/2873/files#diff-33f1852514318cc87b4ec66a5c67372b8161934cc6aa5c89cae7fbb09d2dfadfR1348 to enforce all Merkle values are computed before calling `PopulateMerkleValues`. (read 3. as well) It may or may not happen that all Merkle values are computed.
3. Lazy loading will load each trie node from their Merkle value... so really it will always be there, and that 'caching of Merkle value enforcement' won't be needed anymore. That is also partly why the enforcement is outside this `PopulateMerkleValues` function.